### PR TITLE
remove the memcpy-on-equal-ptrs assumption

### PR DESCRIPTION
**Pull request by RalfJung**
One of the libc we support, musl, [defines `memcpy` with `restrict` pointers](https://git.musl-libc.org/cgit/musl/tree/src/string/memcpy.c#n5). This in fact matches the definition in the C standard. Calling that `memcpy` with overlapping pointers is clearly UB, who knows what the compiler did when optimizing this `memcpy` -- it certainly assumed source and destination to be disjoint.

Lucky enough, it does not seem like we actually need this assumption that `memcpy(p, p, n)` is always allowed. clang and GCC need it since they use `memcpy` to compile C assignments, but [we use memmove for similar code](https://godbolt.org/z/bcW85WYcM). There are no known cases where LLVM introduces calls to memcpy on equal pointers itself. (And if there were, that would be a soundness bug in rustc due to the musl issue mentioned above.)

This does mean we must make sure to never call the LLVM `memcpy` builtin on equal ranges even though the LangRef says that is allowed. Currently that is the case so we just need to make sure it remains the case. :) Cc @rust-lang/opsem @rust-lang/wg-llvm

### CODE REVIEW

**Comment by the8472** [comment]
> There are no known cases where LLVM introduces calls to memcpy on equal pointers itself. 

A black swan has not been observed, therefore all swans are white? Or do we have stronger confirmation that LLVM won't do this?

**Comment by RalfJung** [comment]
I didn't say it's great but I don't know any better way than asking @nikic. :shrug:  LLVM does not even document anywhere what it assumes about the libc `memcpy`, so it's hard to say anything with reasonable amounts of certainty.

That said, the main reason I am proposing this is that if we do see a black swan, then the bit of text we have here doesn't actually help. We are unsound on musl. So I want to have the text removed from the docs to make it clear that this would be a soundness bug.

**Comment by bjorn3** [comment]
In the `StatementKind::Assign` implementation I can only find memcpy, not memmove: https://github.com/rust-lang/rust/blob/b29a1e00f850e548f3021ea523d0e143724fa9b7/compiler/rustc_codegen_ssa/src/mir/operand.rs#L424

**Comment by bjorn3** [comment]
Looks like rustc indeed emits `llvm.memcpy` for your example, but LLVM itself turns it into memmove: https://godbolt.org/z/3PKP47z9a

File: library/core/src/lib.rs [hunk 1]
```diff
@@ -24,9 +24,8 @@
 //!   which are generated by Rust codegen backends. Additionally, this library can make explicit
 //!   calls to `strlen`. Their signatures are the same as found in C, but there are extra
 //!   assumptions about their semantics: For `memcpy`, `memmove`, `memset`, `memcmp`, and `bcmp`, if
-//!   the `n` parameter is 0, the function is assumed to not be UB. Furthermore, for `memcpy`, if
-//!   source and target pointer are equal, the function is assumed to not be UB.
-//!   (Note that these are standard assumptions among compilers:
+//!   the `n` parameter is 0, the function is assumed to not be UB, even if the pointers are NULL or
+//!   dangling. (Note that making extra assumptions about these functions is common among compilers:
```
Comment (cuviper) [inline library/core/src/lib.rs:28]: While removing the "equal" part, you added "NULL or dangling" -- isn't that a *new* assumption? Or are we already living by that elsewhere?

Comment (RalfJung) [reply library/core/src/lib.rs:28]: That's a clarification. We already have "if the n parameter is 0, the function is assumed to not be UB", which was already meant to include *all* pointers, including NULL or dangling. I just made it more explicit.

**Pull request merged**